codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java

```diff
@@ -0,0 +1,304 @@
+            });
+        }
+    }
+
+    /**
+     * Process a single write. Guaranteed to eventually be followed by a {@link #doWriteComplete()},
+     * which denotes the end of the batch of writes. May be called from any thread.
+     */
+    protected abstract void doWrite(Object msg) throws Exception;
+
+    /**
+     * Process end of batch of {@link #doWrite()}s. May be called from any thread.
+     */
+    protected abstract void doWriteComplete();
+
+    /**
+     * The ideal thread for events like {@link #doWrite()} to be processed on. May be used for
+     * efficient batching, but not required.
+     */
+    protected abstract EventExecutor preferredEventExecutor();
+
+    /**
+     * {@code bytes}-count of bytes provided to {@link #fireChildRead} have been read. May be called
+     * from any thread.
+     */
+    protected abstract void bytesConsumed(int bytes);
+
+    protected void fireChildRead(final Object msg) {
+        if (eventLoop().inEventLoop()) {
+            fireChildRead0(msg);
+        } else {
+            eventLoop().execute(new OneTimeTask() {
+                @Override
+                public void run() {
+                    fireChildRead0(msg);
+                }
+            });
+        }
+    }
+
+    private void fireChildRead0(Object msg) {
+        if (closed) {
+            ReferenceCountUtil.release(msg);
+            return;
+        }
+        if (readInProgress) {
+            assert inboundBuffer.isEmpty();
+            // Check for null because inboundBuffer doesn't support null; we want to be consistent
+            // for what values are supported.
+            doRead0(checkNotNull(msg, "msg"), unsafe().recvBufAllocHandle());
+            if (!unsafe().recvBufAllocHandle().continueReading()) {
+                fireChildReadCompleteTask.run();
+            }
+        } else {
+            inboundBuffer.add(msg);
+        }
+    }
+
+    protected void fireChildReadComplete() {
+        if (eventLoop().inEventLoop()) {
+            fireChildReadCompleteTask.run();
+        } else {
+            eventLoop().execute(fireChildReadCompleteTask);
+        }
+    }
+
+    /**
+     * Returns whether reads should continue.
+     */
+    private boolean doRead0(Object msg, RecvByteBufAllocator.Handle allocHandle) {
+        if (msg == CLOSE_MESSAGE) {
+            allocHandle.readComplete();
+            pipeline().fireChannelReadComplete();
+            unsafe().close(newPromise());
+            return false;
+        }
+        int numBytesToBeConsumed = 0;
+        if (msg instanceof Http2DataFrame) {
+            Http2DataFrame data = (Http2DataFrame) msg;
+            numBytesToBeConsumed = data.content().readableBytes() + data.padding();
+            allocHandle.lastBytesRead(numBytesToBeConsumed);
+        } else {
+            // Count other frames in some way. 9 is arbitrary, but is also the minimum size of an
+            // HTTP/2 frame.
+            allocHandle.lastBytesRead(9);
```

Comment:
@ejona86 - I guess it would be up to the frame implementations to decide how to implement. We could preserve the encoded size in the future or for now headers could have a "rough" estimate `9 + headers.size() * 10`. Not a big deal, but seems a bit more natural to let the frames decide instead of guessing here (or having frame specific knowledge here).